executor/src/witgen/expression_evaluator.rs

```diff
@@ -55,47 +55,62 @@ where
         op: &AlgebraicBinaryOperator,
         right: &'a Expression<T>,
     ) -> AffineResult<&'a AlgebraicReference, T> {
-        let left = self.evaluate(left);
-
-        // Short-circuit multiplication by zero.
-        if *op == AlgebraicBinaryOperator::Mul {
-            if let Ok(zero) = &left {
-                if zero.constant_value().map(|z| z.is_zero()) == Some(true) {
-                    return Ok(zero.clone());
+        match op {
+            AlgebraicBinaryOperator::Add => {
+                let left_expr = self.evaluate(left)?;
+                if left_expr.is_zero() {
+                    return self.evaluate(right);
                 }
+                let right_expr = self.evaluate(right)?;
+                if right_expr.is_zero() {
+                    return Ok(left_expr);
+                }
+                Ok(left_expr + right_expr)
             }
-        }
-        let right = self.evaluate(right);
-
-        match (left, op, right) {
-            // Short-circuit multiplication by zero for "right".
-            (_, AlgebraicBinaryOperator::Mul, Ok(zero))
-                if zero.constant_value().map(|z| z.is_zero()) == Some(true) =>
-            {
-                Ok(zero)
-            }
-            (Ok(left), op, Ok(right)) => match op {
-                AlgebraicBinaryOperator::Add => Ok(left + right),
-                AlgebraicBinaryOperator::Sub => Ok(left - right),
-                AlgebraicBinaryOperator::Mul => {
-                    if let Some(f) = left.constant_value() {
-                        Ok(right * f)
-                    } else if let Some(f) = right.constant_value() {
-                        Ok(left * f)
-                    } else {
-                        Err(IncompleteCause::QuadraticTerm)
+            AlgebraicBinaryOperator::Sub => Ok(self.evaluate(left)? - self.evaluate(right)?),
+            AlgebraicBinaryOperator::Mul => {
+                // don't short circuit on err as rhs might still be 0
+                let left_res = self.evaluate(left);
+                match left_res {
+                    Ok(left_expr) if left_expr.is_zero() => Ok(left_expr),
+                    Ok(left_expr) if left_expr.is_one() => self.evaluate(right),
+                    Ok(left_expr) => {
+                        if let Some(n) = left_expr.constant_value() {
+                            return Ok(self.evaluate(right)? * n);
+                        }
+                        // lhs not a constant
+                        let right_expr = self.evaluate(right)?;
```

Comment:
This is what I meant:
```suggestion
                        let right_expr = self.evaluate(right)?;
                        if let Some(n) = left_expr.constant_value() {
                            return Ok(right_expr * n);
                        }
                        // lhs not a constant
```